{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1bfec4af-e675-4c43-810c-869e0602fe99",
   "metadata": {},
   "source": [
    "# Make Synthetic Genome\n",
    "Author : Mathieu Giguere \\\n",
    "Date : 14/08/2024 \\\n",
    "Brief : Strings together the genes of interest of some species into a singular synthetic genome. \\\n",
    "Dependencies : "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "708632c5-f88a-4fe6-a007-16cb5e296d21",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Plan\n",
    "\n",
    "### 1. Read genome files (fasta) -> big string\n",
    "### 2. Read gffs files -> then use regex to extract the coordinates of the genes of interest\n",
    "### 3. For each genome, for each gene of interest, extract its dna sequence in the appropriate genome file\n",
    "### 4. Append each string into 1 genome with spacers\n",
    "### 5. Write synthetic genome file in fasta format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c691f46d-e402-4388-9b6a-7a1083164898",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "117a69c9-2675-475a-a23e-6879522e034e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "species_genome_file_list = [\"fasta_files/unzipped/Aspergillus_fumigatus.ASM265v1.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/C_albicans_SC5314_version_A22-s07-m01-r195_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/C_auris_B8441_version_s01-m03-r08_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/C_parapsilosis_CDC317_version_s01-m06-r03_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/Candida_tropicalis.GCA000006335v3.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/Cryptococcus_neoformans.ASM9104v1.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/C_glabrata_CBS138_version_s05-m03-r06_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/Pichia_kudriavzevii_gca_000764455.ASM76445v1.dna.toplevel.fa\"]\n",
    "\n",
    "species_list = [\"Aspergillus fumigatus\", \"Candida albicans\", \"Candida auris\", \"Candida parapsilosis\", \"Candida tropicalis\", \"Cryptococcus neoformans\", \"Nasakeomyces glabrata\", \"Pichia kudriavzevii\"]\n",
    "\n",
    "gff_files_list = [\"gff_files/Aspergillus_fumigatus.ASM265v1.59.gff3\",\n",
    "                 \"gff_files/C_albicans_SC5314_version_A22-s07-m01-r195.gff\",\n",
    "                 \"gff_files/C_auris_B8441_version_s01-m03-r08.gff\",\n",
    "                 \"gff_files/C_parapsilosis_CDC317_version_s01-m06-r03.gff\",\n",
    "                 \"gff_files/Candida_tropicalis.GCA000006335v3.59.gff3\",\n",
    "                 \"gff_files/Cryptococcus_neoformans.ASM9104v1.59.gff3\",\n",
    "                \"gff_files/C_glabrata_CBS138_version_s05-m03-r06.gff\",\n",
    "                 \"gff_files/Pichia_kudriavzevii_gca_000764455.ASM76445v1.59.gff3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "35c07897-d9f9-41d1-8b40-08764fa8e63a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_candid = [\"gff_files/C_albicans_SC5314_version_A22-s07-m01-r195.gff\",\n",
    "              \"gff_files/C_auris_B8441_version_s01-m03-r08.gff\",\n",
    "              \"gff_files/C_glabrata_CBS138_version_s05-m03-r06.gff\",\n",
    "              \"gff_files/C_parapsilosis_CDC317_version_s01-m06-r03.gff\"]\n",
    "\n",
    "gff_ensembl = [\"gff_files/Aspergillus_fumigatus.ASM265v1.59.gff3\",\n",
    "               \"gff_files/Candida_tropicalis.GCA000006335v3.59.gff3\",\n",
    "               \"gff_files/Cryptococcus_neoformans.ASM9104v1.59.gff3\",\n",
    "               \"gff_files/Pichia_kudriavzevii_gca_000764455.ASM76445v1.59.gff3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "780bb318-12df-472a-ad02-0b6986473073",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aspergillus fumigatus</td>\n",
       "      <td>[AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Candida albicans</td>\n",
       "      <td>[C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Candida auris</td>\n",
       "      <td>[B9J08_001448, B9J08_003737, B9J08_000964, B9J...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>[CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Candida tropicalis</td>\n",
       "      <td>[CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>[CNA00300, CNN02320, CNE02100, CNA05950, CNF04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Nasakeomyces glabrata</td>\n",
       "      <td>[CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pichia kudriavzevii</td>\n",
       "      <td>[JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Species                                            Gene ID\n",
       "0    Aspergillus fumigatus  [AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...\n",
       "1         Candida albicans  [C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...\n",
       "2            Candida auris  [B9J08_001448, B9J08_003737, B9J08_000964, B9J...\n",
       "3     Candida parapsilosis  [CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...\n",
       "4       Candida tropicalis  [CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...\n",
       "5  Cryptococcus neoformans  [CNA00300, CNN02320, CNE02100, CNA05950, CNF04...\n",
       "6    Nasakeomyces glabrata  [CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...\n",
       "7      Pichia kudriavzevii  [JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make dataframe that links the species to their genes of interest\n",
    "\n",
    "df = pd.read_csv(\"gene-species.csv\")\n",
    "\n",
    "df = df.groupby('Species')['Gene ID'].apply(list).reset_index()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7075ff7d-e8a4-4204-b289-da02deb5da84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasta_files/unzipped/Aspergillus_fumigatus.ASM265v1.dna.toplevel.fa\n",
      "fasta_files/unzipped/C_albicans_SC5314_version_A22-s07-m01-r195_chromosomes.fasta\n",
      "fasta_files/unzipped/C_auris_B8441_version_s01-m03-r08_chromosomes.fasta\n",
      "fasta_files/unzipped/C_parapsilosis_CDC317_version_s01-m06-r03_chromosomes.fasta\n",
      "fasta_files/unzipped/Candida_tropicalis.GCA000006335v3.dna.toplevel.fa\n",
      "fasta_files/unzipped/Cryptococcus_neoformans.ASM9104v1.dna.toplevel.fa\n",
      "fasta_files/unzipped/C_glabrata_CBS138_version_s05-m03-r06_chromosomes.fasta\n",
      "fasta_files/unzipped/Pichia_kudriavzevii_gca_000764455.ASM76445v1.dna.toplevel.fa\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Landmarks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aspergillus fumigatus</td>\n",
       "      <td>[AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...</td>\n",
       "      <td>{'1': 0, '2': 4918979, '3': 9763451, '4': 1384...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Candida albicans</td>\n",
       "      <td>[C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...</td>\n",
       "      <td>{'Ca22chr1A_C_albicans_SC5314': 0, 'Ca22chr1B_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Candida auris</td>\n",
       "      <td>[B9J08_001448, B9J08_003737, B9J08_000964, B9J...</td>\n",
       "      <td>{'PEKT02000001_C_auris_B8441': 0, 'PEKT0200000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>[CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...</td>\n",
       "      <td>{'Contig005504_C_parapsilosis_CDC317': 0, 'Con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Candida tropicalis</td>\n",
       "      <td>[CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...</td>\n",
       "      <td>{'GG692418': 0, 'GG692395': 50304, 'GG692396':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>[CNA00300, CNN02320, CNE02100, CNA05950, CNF04...</td>\n",
       "      <td>{'1': 0, '2': 2300533, '3': 3932840, '4': 6038...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Nasakeomyces glabrata</td>\n",
       "      <td>[CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...</td>\n",
       "      <td>{'ChrA_C_glabrata_CBS138': 0, 'ChrB_C_glabrata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pichia kudriavzevii</td>\n",
       "      <td>[JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200...</td>\n",
       "      <td>{'scaffold00001': 0, 'scaffold00002': 607893, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Species                                            Gene ID  \\\n",
       "0    Aspergillus fumigatus  [AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...   \n",
       "1         Candida albicans  [C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...   \n",
       "2            Candida auris  [B9J08_001448, B9J08_003737, B9J08_000964, B9J...   \n",
       "3     Candida parapsilosis  [CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...   \n",
       "4       Candida tropicalis  [CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...   \n",
       "5  Cryptococcus neoformans  [CNA00300, CNN02320, CNE02100, CNA05950, CNF04...   \n",
       "6    Nasakeomyces glabrata  [CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...   \n",
       "7      Pichia kudriavzevii  [JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200...   \n",
       "\n",
       "                                           Landmarks  \n",
       "0  {'1': 0, '2': 4918979, '3': 9763451, '4': 1384...  \n",
       "1  {'Ca22chr1A_C_albicans_SC5314': 0, 'Ca22chr1B_...  \n",
       "2  {'PEKT02000001_C_auris_B8441': 0, 'PEKT0200000...  \n",
       "3  {'Contig005504_C_parapsilosis_CDC317': 0, 'Con...  \n",
       "4  {'GG692418': 0, 'GG692395': 50304, 'GG692396':...  \n",
       "5  {'1': 0, '2': 2300533, '3': 3932840, '4': 6038...  \n",
       "6  {'ChrA_C_glabrata_CBS138': 0, 'ChrB_C_glabrata...  \n",
       "7  {'scaffold00001': 0, 'scaffold00002': 607893, ...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get landmarks !\n",
    "\n",
    "list1 = []\n",
    "\n",
    "for s, gff_file in enumerate(gff_files_list):\n",
    "    \n",
    "    with open(gff_file) as gff:\n",
    "        gff_content = gff.read()\n",
    "\n",
    "    landmark_dict = {}\n",
    "    landmark = 0\n",
    "    coord_list = []\n",
    "    genome_file = species_genome_file_list[s]\n",
    "    print(genome_file)\n",
    "    \n",
    "    if gff_file in gff_ensembl:\n",
    "        \n",
    "        ## \n",
    "        with open(genome_file) as gen:\n",
    "            gen_content = gen.read()\n",
    "        \n",
    "        list_ids = []\n",
    "        ids_tofind = re.findall(\">.*? \", gen_content)\n",
    "        \n",
    "        for identificator_index in range(len(ids_tofind)):\n",
    "            iden = re.sub(\">\", \"\", ids_tofind[identificator_index])\n",
    "            iden = re.sub(\" \", \"\", iden)\n",
    "            list_ids.append(iden)\n",
    "        \n",
    "        #print(ids_tofind)\n",
    "        #print(list_ids)\n",
    "        \n",
    "        #list_gff_ids = []\n",
    "        for iiindex, identificator in enumerate(list_ids):\n",
    "            landma_line = re.findall(f\"##sequence-region   {identificator}.*\\\\n\", gff_content)[0]\n",
    "            #print(landma_line)\n",
    "            landmarc = re.sub(\".* \", \"\", landma_line)\n",
    "            #print(landmarc)\n",
    "            #coord_list.append(landmarc)\n",
    "        \n",
    "        #######\n",
    "        #gff_landmarks = re.findall(\"##sequence-region.*\\\\n\", gff_content)\n",
    "        \n",
    "        #print(gff_landmarks)\n",
    "        \n",
    "        #for i, ld in enumerate(gff_landmarks):\n",
    "        #    index = re.findall(\"  .+? \", ld)[0]\n",
    "        #    index = re.sub(\" \", \"\", index)\n",
    "        #    coord = re.sub(\".*? \", \"\", ld)\n",
    "        #    \n",
    "            landmark += int(landmarc)\n",
    "            coord_list.append(landmark)\n",
    "            \n",
    "            if iiindex != 0:\n",
    "                landmark_dict[identificator] = coord_list[iiindex-1]\n",
    "            else:\n",
    "                landmark_dict[identificator] = 0\n",
    "\n",
    "        list1.append(landmark_dict)\n",
    "        \n",
    "        \n",
    "        \n",
    "        ##\n",
    "        \n",
    "    else:\n",
    "        gff_landmarks = re.findall(r\"^.*CGD\\tchromosome.*|^.*CGD\\tcontig.*\", gff_content, flags=re.MULTILINE)\n",
    "        \n",
    "        for i, ld in enumerate(gff_landmarks):\n",
    "            index = re.findall(r\"(?:^|(?:[.!?]\\s))(\\w+)\", ld)[0]\n",
    "            coord = re.findall(\"\\t[0-9]{4}.+?\\t\", ld)[0]\n",
    "            coord = re.sub(\"[^0-9]\", \"\", coord)\n",
    "\n",
    "            landmark += int(coord)\n",
    "            coord_list.append(landmark)\n",
    "            \n",
    "            if i != 0:\n",
    "                landmark_dict[index] = coord_list[i-1]\n",
    "            else:\n",
    "                landmark_dict[index] = 0\n",
    "        \n",
    "        list1.append(landmark_dict)\n",
    "\n",
    "\n",
    "df[\"Landmarks\"] = list1 \n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "530428ab-764a-462d-b76a-0bddb2f562c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aspergillus_fumigatus\n",
      "Candida_albicans\n",
      "Candida_auris\n",
      "Candida_parapsilosis\n",
      "Candida_tropicalis\n",
      "Cryptococcus_neoformans\n",
      "Nasakeomyces_glabrata\n",
      "Pichia_kudriavzevii\n"
     ]
    }
   ],
   "source": [
    "synth = \"\"\n",
    "spacer = \"NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\"\n",
    "\n",
    "for i, file in enumerate(species_genome_file_list):\n",
    "    species = species_list[i]\n",
    "    #print(species)\n",
    "    sp = re.sub(\" \", \"_\", species)\n",
    "    print(sp)\n",
    "    synth += f\">{sp}\\n\"\n",
    "    \n",
    "    with open(file) as genome:\n",
    "        genome_content = genome.read()\n",
    "    \n",
    "    # Removes first line puts a space\n",
    "    dna = re.sub(\">.*\", \"\", genome_content)\n",
    "    dna = re.sub(\"\\n\", \"\", dna)\n",
    "    \n",
    "    gff_file = gff_files_list[i]\n",
    "    with open(gff_file) as gff:\n",
    "        gff_content = gff.read()\n",
    "    \n",
    "    gene_list = df['Gene ID'][df['Species'] == species].values[0]\n",
    "    \n",
    "    for g in gene_list:\n",
    "        #print(f'Gene ID : {g}')\n",
    "        \n",
    "        # find gene coordinates\n",
    "        #print(g)\n",
    "        y = re.findall(f\"^.*\\t.*\\tgene\\t.*?{g}\", gff_content, flags=re.MULTILINE)\n",
    "        coordinates = re.findall(\"[^\\\\t]+\", y[0])\n",
    "        #print(f'Coordinates : {coordinates}')\n",
    "        index = coordinates[0]\n",
    "        \n",
    "        thelandmark = df['Landmarks'][df['Species'] == species].values[0][index]\n",
    "        #print(thelandmark)\n",
    "        begin = int(coordinates[3]) + int(thelandmark) -1 # Minus 1 to get to base 0 \n",
    "        end = int(coordinates[4]) + int(thelandmark) # Here we do not substract because we want to include the last nucleotide.\n",
    "        gene_seq = dna[begin:end]\n",
    "        #print(f\"{begin} --- {end}\")\n",
    "        \n",
    "        #print(f'Gene length : {len(gene_seq)}')\n",
    "        #print(gene_seq)\n",
    "        synth += gene_seq + spacer * 5\n",
    "    \n",
    "    synth += '\\n'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83ed85de-55cf-41d8-a44b-e4a3fce00d58",
   "metadata": {},
   "source": [
    "# Fasta file creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd0f411c-4355-4fd3-b144-2a233ea3dda2",
   "metadata": {},
   "outputs": [],
   "source": [
    "synth_w_newlines = re.sub(\"([A-Za-z]{60})\", \"\\\\1\\n\", synth, 0, re.DOTALL)\n",
    "\n",
    "#fasta_header = \">synthetic genome, author GiguereM\\n\"\n",
    "\n",
    "my_synth_genome = open(\"v4_synthetic_genome.fa\", \"w\")\n",
    "my_synth_genome.write(synth_w_newlines)\n",
    "my_synth_genome.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62ec82e9-157e-4eb3-be84-17bcc3cf4df1",
   "metadata": {},
   "source": [
    "# Update !!\n",
    "\n",
    "### Added a header for each species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f07dced3-4e1c-48f7-a408-be1708cfdcae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
