{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c0e304a6-dc12-4a8a-a292-5ee603101319",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from BCBio import GFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "877acb2f-0bcd-4f34-98a0-fa40102c78f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_genome_file_list = [\"fasta_files/unzipped/Aspergillus_fumigatus.ASM265v1.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/C_albicans_SC5314_version_A22-s07-m01-r195_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/C_auris_B8441_version_s01-m03-r08_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/C_glabrata_CBS138_version_s05-m03-r06_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/C_parapsilosis_CDC317_version_s01-m06-r03_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/Candida_tropicalis.GCA000006335v3.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/Cryptococcus_neoformans.ASM9104v1.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/Pichia_kudriavzevii_gca_000764455.ASM76445v1.dna.toplevel.fa\"]\n",
    "\n",
    "species_list = [\"Aspergillus fumigatus\", \"Candida albicans\", \"Candida auris\", \"Nasakeomyces glabrata\", \"Candida parapsilosis\", \"Candida tropicalis\", \"Cryptococcus neoformans\", \"Pichia kudriavzevii\"]\n",
    "\n",
    "gff_files_list = [\"gff_files/Aspergillus_fumigatus.ASM265v1.59.gff3\",\n",
    "                 \"gff_files/C_albicans_SC5314_version_A22-s07-m01-r195.gff\",\n",
    "                 \"gff_files/C_auris_B8441_version_s01-m03-r08.gff\",\n",
    "                 \"gff_files/C_parapsilosis_CDC317_version_s01-m06-r03.gff\",\n",
    "                 \"gff_files/Candida_tropicalis.GCA000006335v3.59.gff3\",\n",
    "                 \"gff_files/Cryptococcus_neoformans.ASM9104v1.59.gff3\",\n",
    "                  \"gff_files/C_glabrata_CBS138_version_s05-m03-r06.gff\",\n",
    "                 \"gff_files/Pichia_kudriavzevii_gca_000764455.ASM76445v1.59.gff3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "8804bd91-e387-4772-8c30-e5972cc9d3e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aspergillus fumigatus</td>\n",
       "      <td>[AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Candida albicans</td>\n",
       "      <td>[C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Candida auris</td>\n",
       "      <td>[B9J08_001448, B9J08_003737, B9J08_000964, B9J...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>[CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Candida tropicalis</td>\n",
       "      <td>[CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>[CNA00300, CNN02320, CNE02100, CNA05950, CNF04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Nasakeomyces glabrata</td>\n",
       "      <td>[CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pichia kudriavzevii</td>\n",
       "      <td>[JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Species                                            Gene ID\n",
       "0    Aspergillus fumigatus  [AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...\n",
       "1         Candida albicans  [C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...\n",
       "2            Candida auris  [B9J08_001448, B9J08_003737, B9J08_000964, B9J...\n",
       "3     Candida parapsilosis  [CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...\n",
       "4       Candida tropicalis  [CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...\n",
       "5  Cryptococcus neoformans  [CNA00300, CNN02320, CNE02100, CNA05950, CNF04...\n",
       "6    Nasakeomyces glabrata  [CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...\n",
       "7      Pichia kudriavzevii  [JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200..."
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make dataframe that links the species to their genes of interest\n",
    "\n",
    "df = pd.read_csv(\"gene-species.csv\")\n",
    "\n",
    "df = df.groupby('Species')['Gene ID'].apply(list).reset_index()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "0dae429f-683c-4d24-961d-d4ac52078038",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_candid = [\"gff_files/C_albicans_SC5314_version_A22-s07-m01-r195.gff\",\n",
    "              \"gff_files/C_auris_B8441_version_s01-m03-r08.gff\",\n",
    "              \"gff_files/C_glabrata_CBS138_version_s05-m03-r06.gff\",\n",
    "              \"gff_files/C_parapsilosis_CDC317_version_s01-m06-r03.gff\"]\n",
    "\n",
    "gff_ensembl = [\"gff_files/Aspergillus_fumigatus.ASM265v1.59.gff3\",\n",
    "               \"gff_files/Candida_tropicalis.GCA000006335v3.59.gff3\",\n",
    "               \"gff_files/Cryptococcus_neoformans.ASM9104v1.59.gff3\",\n",
    "               \"gff_files/Pichia_kudriavzevii_gca_000764455.ASM76445v1.59.gff3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "2d608fd0-b652-49fb-b95f-4473931528d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Landmarks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aspergillus fumigatus</td>\n",
       "      <td>[AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...</td>\n",
       "      <td>{'1': 4918978, '2': 9763449, '3': 13842615, '4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Candida albicans</td>\n",
       "      <td>[C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...</td>\n",
       "      <td>{'Ca22chr1A_C_albicans_SC5314': 3188340, 'Ca22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Candida auris</td>\n",
       "      <td>[B9J08_001448, B9J08_003737, B9J08_000964, B9J...</td>\n",
       "      <td>{'PEKT02000001_C_auris_B8441': 1083521, 'PEKT0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>[CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...</td>\n",
       "      <td>{'Contig005504_C_parapsilosis_CDC317': 898304,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Candida tropicalis</td>\n",
       "      <td>[CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...</td>\n",
       "      <td>{'GG692395': 2474447, 'GG692396': 4783116, 'GG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>[CNA00300, CNN02320, CNE02100, CNA05950, CNF04...</td>\n",
       "      <td>{'1': 2300532, '10': 3386251, '11': 4406096, '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Nasakeomyces glabrata</td>\n",
       "      <td>[CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...</td>\n",
       "      <td>{'ChrA_C_glabrata_CBS138': 527799, 'ChrB_C_gla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pichia kudriavzevii</td>\n",
       "      <td>[JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200...</td>\n",
       "      <td>{'contig03002': 1995, 'contig03003': 3981, 'co...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Species                                            Gene ID  \\\n",
       "0    Aspergillus fumigatus  [AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...   \n",
       "1         Candida albicans  [C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...   \n",
       "2            Candida auris  [B9J08_001448, B9J08_003737, B9J08_000964, B9J...   \n",
       "3     Candida parapsilosis  [CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...   \n",
       "4       Candida tropicalis  [CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...   \n",
       "5  Cryptococcus neoformans  [CNA00300, CNN02320, CNE02100, CNA05950, CNF04...   \n",
       "6    Nasakeomyces glabrata  [CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...   \n",
       "7      Pichia kudriavzevii  [JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200...   \n",
       "\n",
       "                                           Landmarks  \n",
       "0  {'1': 4918978, '2': 9763449, '3': 13842615, '4...  \n",
       "1  {'Ca22chr1A_C_albicans_SC5314': 3188340, 'Ca22...  \n",
       "2  {'PEKT02000001_C_auris_B8441': 1083521, 'PEKT0...  \n",
       "3  {'Contig005504_C_parapsilosis_CDC317': 898304,...  \n",
       "4  {'GG692395': 2474447, 'GG692396': 4783116, 'GG...  \n",
       "5  {'1': 2300532, '10': 3386251, '11': 4406096, '...  \n",
       "6  {'ChrA_C_glabrata_CBS138': 527799, 'ChrB_C_gla...  \n",
       "7  {'contig03002': 1995, 'contig03003': 3981, 'co...  "
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list1 = []\n",
    "\n",
    "for gff_file in gff_files_list:\n",
    "    \n",
    "    with open(gff_file) as gff:\n",
    "        gff_content = gff.read()\n",
    "\n",
    "    landmark_dict = {}\n",
    "    landmark = 0\n",
    "    \n",
    "    if gff_file in gff_ensembl:\n",
    "        gff_landmarks = re.findall(\"##sequence-region.*\\\\n\", gff_content)\n",
    "        \n",
    "        for i in gff_landmarks:\n",
    "            index = re.findall(\"  .+? \", i)[0]\n",
    "            index = re.sub(\" \", \"\", index)\n",
    "            coord = re.sub(\".*? \", \"\", i)\n",
    "            landmark += int(coord) - 1\n",
    "            landmark_dict[index] = landmark\n",
    "        \n",
    "        list1.append(landmark_dict)\n",
    "        \n",
    "    else:\n",
    "        gff_landmarks = re.findall(r\"^.*CGD\\tchromosome.*|^.*CGD\\tcontig.*\", gff_content, flags=re.MULTILINE)\n",
    "        for i in gff_landmarks:\n",
    "            #print(i)\n",
    "            index = re.findall(r\"(?:^|(?:[.!?]\\s))(\\w+)\", i)[0]\n",
    "            coord = re.findall(\"\\t[0-9]{4}.+?\\t\", i)[0]\n",
    "            coord = re.sub(\"[^0-9]\", \"\", coord)\n",
    "            #print(coord)\n",
    "            landmark += int(coord) - 1\n",
    "            landmark_dict[index] = landmark\n",
    "\n",
    "        list1.append(landmark_dict)\n",
    "    \n",
    "#print(list1)\n",
    "\n",
    "df[\"Landmarks\"] = list1 \n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "884fcd9a-b315-472d-8e6a-315835dd7862",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ChrA_C_glabrata_CBS138': 527799,\n",
       " 'ChrB_C_glabrata_CBS138': 1040453,\n",
       " 'ChrC_C_glabrata_CBS138': 1634978,\n",
       " 'ChrD_C_glabrata_CBS138': 2302877,\n",
       " 'ChrE_C_glabrata_CBS138': 3010299,\n",
       " 'ChrF_C_glabrata_CBS138': 3967813,\n",
       " 'ChrG_C_glabrata_CBS138': 4978104,\n",
       " 'ChrH_C_glabrata_CBS138': 6035491,\n",
       " 'ChrI_C_glabrata_CBS138': 7178015,\n",
       " 'ChrJ_C_glabrata_CBS138': 8425577,\n",
       " 'ChrK_C_glabrata_CBS138': 9732666,\n",
       " 'ChrL_C_glabrata_CBS138': 11260929,\n",
       " 'ChrM_C_glabrata_CBS138': 12726200,\n",
       " 'mito_C_glabrata_CBS138': 12746262}"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Landmarks'][df['Species'] == 'Nasakeomyces glabrata'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6e722469-cdb2-4d7b-b09f-3eab5f9d46c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aspergillus fumigatus\n",
      "AFUA_1G05050\n",
      "[1449701:1450353](+)\n",
      "1\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "AFUA_1G10910\n",
      "[2848154:2850137](-)\n",
      "1\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "AFUA_2G00320\n",
      "[61019:62153](+)\n",
      "2\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "AFUA_2G03700\n",
      "[984933:988503](+)\n",
      "2\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "AFUA_2G15130\n",
      "[3997635:4002389](-)\n",
      "2\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "AFUA_3G05760\n",
      "[1406208:1407279](-)\n",
      "3\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "AFUA_4G06890\n",
      "[1780203:1781822](-)\n",
      "4\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "AFUA_4G12560\n",
      "[3289800:3292477](-)\n",
      "4\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "AFUA_5G05460\n",
      "[1268052:1268981](-)\n",
      "5\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "AFUA_5G06070\n",
      "[1446972:1451166](+)\n",
      "5\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "AFUA_5G10370\n",
      "[2654820:2655836](-)\n",
      "5\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "AFUA_6G05140\n",
      "[1228801:1229934](-)\n",
      "6\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "AFUA_6G12400\n",
      "[3126473:3132292](-)\n",
      "6\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "AFUA_7G03740\n",
      "[804300:806031](-)\n",
      "7\n",
      "[('1', 0, 4918979), ('2', 0, 4844472), ('3', 0, 4079167), ('4', 0, 3923705), ('5', 0, 3948441), ('6', 0, 3778736), ('7', 0, 2058334), ('8', 0, 1833124)]\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "hello\n",
      "Candida albicans\n",
      "C1_00710C_A\n",
      "[123868:125597](-)\n",
      "Ca22chr1A_C_albicans_SC5314\n",
      "None\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: 'Ca22chr1A_C_albicans_SC5314'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[49], line 30\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[38;5;28mprint\u001b[39m(rec\u001b[38;5;241m.\u001b[39mid)\n\u001b[1;32m     29\u001b[0m \u001b[38;5;28mprint\u001b[39m(rec\u001b[38;5;241m.\u001b[39mannotations\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msequence-region\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[0;32m---> 30\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m n \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m0\u001b[39m, \u001b[38;5;28;43mint\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mrec\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mid\u001b[49m\u001b[43m)\u001b[49m):\n\u001b[1;32m     31\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhello\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     32\u001b[0m \u001b[38;5;66;03m#print(rec.annotations.get(\"sequence-region\")[0][2])\u001b[39;00m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;66;03m#print(feature.location.start)\u001b[39;00m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;66;03m#print(feature.location.end)\u001b[39;00m\n\u001b[1;32m     35\u001b[0m \u001b[38;5;66;03m#print(feature.coordinates)\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'Ca22chr1A_C_albicans_SC5314'"
     ]
    }
   ],
   "source": [
    "for i, file in enumerate(species_genome_file_list):\n",
    "    species = species_list[i]\n",
    "    print(species)\n",
    "    with open(file) as genome:\n",
    "        genome_content = genome.read()\n",
    "    \n",
    "    # Removes first line puts a space\n",
    "    dna = re.sub(\">.*\", \"\", genome_content)\n",
    "    dna = re.sub(\"\\n\", \"\", dna)\n",
    "\n",
    "    gff_file = gff_files_list[i]\n",
    "    in_handle = open(gff_file)\n",
    "\n",
    "    gene_list = df['Gene ID'][df['Species'] == species].values[0]\n",
    "\n",
    "    for rec in GFF.parse(in_handle):\n",
    "        #print(rec.id)\n",
    "        #print(rec)\n",
    "        for feature in rec.features:\n",
    "            #print(feature.type)\n",
    "            if feature.type == \"gene\":\n",
    "                #print(feature.qualifiers)\n",
    "                id_gene = re.sub(\".*:\", \"\", feature.qualifiers.get(\"ID\")[0])\n",
    "                #print(id_gene)\n",
    "                if id_gene in gene_list:\n",
    "                    print(id_gene)\n",
    "                    print(feature.location)\n",
    "                    print(rec.id)\n",
    "                    print(rec.annotations.get(\"sequence-region\"))\n",
    "                    for n in range(0, int(rec.id)):\n",
    "                        print(\"hello\")\n",
    "                    #print(rec.annotations.get(\"sequence-region\")[0][2])\n",
    "                    #print(feature.location.start)\n",
    "                    #print(feature.location.end)\n",
    "                    #print(feature.coordinates)\n",
    "\n",
    "    in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "50dd9c8f-82db-481f-beb8-3b0dad4f6d92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C1_00710C_A\n",
      "[123868:125597](-)\n",
      "ID: Ca22chr1A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1464\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 3188341\n",
      "C1_00800C_A\n",
      "[154632:155286](-)\n",
      "ID: Ca22chr1A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1464\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 3188341\n",
      "C1_02420C_A\n",
      "[505968:511662](-)\n",
      "ID: Ca22chr1A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1464\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 3188341\n",
      "C1_03780C_A\n",
      "[790363:793585](-)\n",
      "ID: Ca22chr1A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1464\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 3188341\n",
      "C1_04770C_A\n",
      "[991621:992782](-)\n",
      "ID: Ca22chr1A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1464\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 3188341\n",
      "C1_08460C_A\n",
      "[1858135:1860274](-)\n",
      "ID: Ca22chr1A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1464\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 3188341\n",
      "C1_08590C_A\n",
      "[1880546:1882037](-)\n",
      "ID: Ca22chr1A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1464\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 3188341\n",
      "C3_02220W_A\n",
      "[478900:480400](+)\n",
      "ID: Ca22chr3A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 799\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1799298\n",
      "C3_04890W_A\n",
      "[1047886:1052386](+)\n",
      "ID: Ca22chr3A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 799\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1799298\n",
      "C3_05220W_A\n",
      "[1146044:1150550](+)\n",
      "ID: Ca22chr3A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 799\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1799298\n",
      "C3_05920W_A\n",
      "[1329262:1332589](+)\n",
      "ID: Ca22chr3A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 799\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1799298\n",
      "C3_06850W_A\n",
      "[1568450:1570004](+)\n",
      "ID: Ca22chr3A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 799\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1799298\n",
      "C3_07860C_A\n",
      "[1778414:1780547](-)\n",
      "ID: Ca22chr3A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 799\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1799298\n",
      "C5_00660C_A\n",
      "[148114:149701](-)\n",
      "ID: Ca22chr5A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 552\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1190869\n",
      "C5_01840C_A\n",
      "[416399:419345](-)\n",
      "ID: Ca22chr5A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 552\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1190869\n",
      "C5_03390C_A\n",
      "[766243:766900](-)\n",
      "ID: Ca22chr5A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 552\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1190869\n",
      "C6_00620W_A\n",
      "[112251:112774](+)\n",
      "ID: Ca22chr6A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 467\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1033292\n",
      "C6_03170C_A\n",
      "[666382:668077](-)\n",
      "ID: Ca22chr6A_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 467\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 1033292\n",
      "CR_00850C_A\n",
      "[192171:197094](-)\n",
      "ID: Ca22chrRA_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1088\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 2286237\n",
      "CR_08780W_A\n",
      "[1893022:1893161](+)\n",
      "ID: Ca22chrRA_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1088\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 2286237\n",
      "CR_08800W_A\n",
      "[1893318:1893469](+)\n",
      "ID: Ca22chrRA_C_albicans_SC5314\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1088\n",
      "/gff-version=['3']\n",
      "Undefined sequence of length 2286237\n"
     ]
    }
   ],
   "source": [
    "in_file = \"gff_files/C_albicans_SC5314_version_A22-s07-m01-r195.gff\"\n",
    "\n",
    "in_handle = open(in_file)\n",
    "\n",
    "species = \"Candida albicans\"\n",
    "\n",
    "gene_list = df['Gene ID'][df['Species'] == species].values[0]\n",
    "\n",
    "for rec in GFF.parse(in_handle):\n",
    "    #print(rec.id)\n",
    "    #print(rec)\n",
    "    for feature in rec.features:\n",
    "        #print(feature.type)\n",
    "        if feature.type == \"gene\":\n",
    "            #print(feature.qualifiers)\n",
    "            id_gene = re.sub(\".*:\", \"\", feature.qualifiers.get(\"ID\")[0])\n",
    "            #print(id_gene)\n",
    "            if id_gene in gene_list:\n",
    "                print(id_gene)\n",
    "                print(feature.location)\n",
    "                print(rec)\n",
    "                #print(feature.location.start)\n",
    "                #print(feature.location.end)\n",
    "                #print(feature.coordinates)\n",
    "\n",
    "in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d3e0245b-1402-4381-80fb-745da5baa8f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seems to have a minus 1 position... so might be in 0 base."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f6eabe28-3d7f-45d7-94a4-55bcbda82f80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DiscoGFFParser',\n",
       " 'GFF3Writer',\n",
       " 'GFFExaminer',\n",
       " 'GFFOutput',\n",
       " 'GFFParser',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__',\n",
       " '__version__',\n",
       " 'parse',\n",
       " 'parse_simple',\n",
       " 'write']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(GFF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e3511ad2-9b8b-442f-904f-9e03f3d06c31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gff_id': {('1',): 12998,\n",
      "            ('2',): 12919,\n",
      "            ('3',): 10734,\n",
      "            ('4',): 9769,\n",
      "            ('5',): 10864,\n",
      "            ('6',): 9985,\n",
      "            ('7',): 4790,\n",
      "            ('8',): 4752},\n",
      " 'gff_source': {('ASM265v1',): 8,\n",
      "                ('ena',): 76439,\n",
      "                ('ena_assembly_gap',): 11,\n",
      "                ('ena_mobile_element',): 353},\n",
      " 'gff_source_type': {('ASM265v1', 'chromosome'): 8,\n",
      "                     ('ena', 'CDS'): 28229,\n",
      "                     ('ena', 'exon'): 28488,\n",
      "                     ('ena', 'five_prime_UTR'): 4,\n",
      "                     ('ena', 'gene'): 9623,\n",
      "                     ('ena', 'mRNA'): 9623,\n",
      "                     ('ena', 'ncRNA_gene'): 229,\n",
      "                     ('ena', 'pseudogene'): 7,\n",
      "                     ('ena', 'pseudogenic_transcript'): 7,\n",
      "                     ('ena', 'tRNA'): 229,\n",
      "                     ('ena_assembly_gap', 'biological_region'): 11,\n",
      "                     ('ena_mobile_element', 'biological_region'): 353},\n",
      " 'gff_type': {('CDS',): 28229,\n",
      "              ('biological_region',): 364,\n",
      "              ('chromosome',): 8,\n",
      "              ('exon',): 28488,\n",
      "              ('five_prime_UTR',): 4,\n",
      "              ('gene',): 9623,\n",
      "              ('mRNA',): 9623,\n",
      "              ('ncRNA_gene',): 229,\n",
      "              ('pseudogene',): 7,\n",
      "              ('pseudogenic_transcript',): 7,\n",
      "              ('tRNA',): 229}}\n"
     ]
    }
   ],
   "source": [
    "import pprint\n",
    "from BCBio.GFF import GFFExaminer\n",
    "\n",
    "in_file = \"gff_files/Aspergillus_fumigatus.ASM265v1.59.gff3\"\n",
    "examiner = GFFExaminer()\n",
    "in_handle = open(in_file)\n",
    "pprint.pprint(examiner.available_limits(in_handle))\n",
    "in_handle.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "43d5a35b-9312-4345-9ea3-3b84b96330b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "298411"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "with open(\"my_synthetic_genome.txt\") as genome:\n",
    "    genome_content = genome.read()\n",
    "\n",
    "# Removes first line puts a space\n",
    "dna = re.sub(\">.*\", \"\", genome_content)\n",
    "dna = re.sub(\"\\n\", \"\", dna)\n",
    "\n",
    "len(dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9cd8615-0642-42df-9783-7f5a7f1817d6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
