{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1bfec4af-e675-4c43-810c-869e0602fe99",
   "metadata": {},
   "source": [
    "# Make Synthetic Genome\n",
    "Author : Mathieu Giguere \\\n",
    "Date : 02/07/2024 \\\n",
    "Brief : Strings together the genes of interest of some species into a singular synthetic genome. \\\n",
    "Dependencies : "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "708632c5-f88a-4fe6-a007-16cb5e296d21",
   "metadata": {},
   "source": [
    "## plan\n",
    "\n",
    "### 1. Lire les fichiers genomes (fasta) -> big string\n",
    "### 2. Lire les fichiers gffs -> Extraire les coordonnees des genes d'interets\n",
    "### 3. Pour chaque genome, chercher la sous-liste des bases de chaque gene d'interet et les appends avec un spacer 'NNNNNNNNNN'\n",
    "### 4. Append tous les str en 1 genome synthetique avec spacer 'N+' adequat\n",
    "### 5. Write synthetic genome file in fasta format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c691f46d-e402-4388-9b6a-7a1083164898",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8213a897-5222-493c-8041-e44ed3582141",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"genes-species.xlsx\") as xlsx:\n",
    "    xlsx_content = xlsx.read()\n",
    "\n",
    "csv = re.sub(\"\\\\t\", \",\", xlsx_content)\n",
    "\n",
    "\n",
    "new_csv = open(\"gene-species.csv\", \"w\")\n",
    "new_csv.write(csv)\n",
    "new_csv.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "780bb318-12df-472a-ad02-0b6986473073",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aspergillus fumigatus</td>\n",
       "      <td>[AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Candida albicans</td>\n",
       "      <td>[C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Candida auris</td>\n",
       "      <td>[B9J08_001448, B9J08_003737, B9J08_000964, B9J...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Candida parapsilosis</td>\n",
       "      <td>[CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Candida tropicalis</td>\n",
       "      <td>[CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>[CNA00300, CNN02320, CNE02100, CNA05950, CNF04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Nasakeomyces glabrata</td>\n",
       "      <td>[CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pichia kudriavzevii</td>\n",
       "      <td>[JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Species                                            Gene ID\n",
       "0    Aspergillus fumigatus  [AFUA_4G06890, AFUA_6G05140, AFUA_2G00320, AFU...\n",
       "1         Candida albicans  [C5_00660C_A, C1_04770C_A, C1_02420C_A, CR_008...\n",
       "2            Candida auris  [B9J08_001448, B9J08_003737, B9J08_000964, B9J...\n",
       "3     Candida parapsilosis  [CPAR2_303740, CPAR2_105550, CPAR2_106400, CPA...\n",
       "4       Candida tropicalis  [CTRG_05283, CTRG_04480, CTRG_04661, CTRG_0268...\n",
       "5  Cryptococcus neoformans  [CNA00300, CNN02320, CNE02100, CNA05950, CNF04...\n",
       "6    Nasakeomyces glabrata  [CAGL0E04334g, CAGL0F01793g, CAGL0G01034g, CAG...\n",
       "7      Pichia kudriavzevii  [JL09_g2508, JL09_g3074, JL09_g1956, JL09_g200..."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"gene-species.csv\")\n",
    "\n",
    "df = df.groupby('Species')['Gene ID'].apply(list).reset_index()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "117a69c9-2675-475a-a23e-6879522e034e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "species_genome_file_list = [\"fasta_files/unzipped/Aspergillus_fumigatus.ASM265v1.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/C_albicans_SC5314_version_A22-s07-m01-r195_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/C_auris_B8441_version_s01-m03-r08_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/C_glabrata_CBS138_version_s05-m03-r06_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/C_parapsilosis_CDC317_version_s01-m06-r03_chromosomes.fasta\",\n",
    "                           \"fasta_files/unzipped/Candida_tropicalis.GCA000006335v3.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/Cryptococcus_neoformans.ASM9104v1.dna.toplevel.fa\",\n",
    "                           \"fasta_files/unzipped/Pichia_kudriavzevii_gca_000764455.ASM76445v1.dna.toplevel.fa\"]\n",
    "\n",
    "species_list = [\"Aspergillus fumigatus\", \"Candida albicans\", \"Candida auris\", \"Nasakeomyces glabrata\", \"Candida parapsilosis\", \"Candida tropicalis\", \"Cryptococcus neoformans\", \"Pichia kudriavzevii\"]\n",
    "\n",
    "gff_files_list = [\"gff_files/Aspergillus_fumigatus.ASM265v1.59.gff3\",\n",
    "                 \"gff_files/C_albicans_SC5314_version_A22-s07-m01-r195.gff\",\n",
    "                 \"gff_files/C_auris_B8441_version_s01-m03-r08.gff\",\n",
    "                 \"gff_files/C_glabrata_CBS138_version_s05-m03-r06.gff\",\n",
    "                 \"gff_files/C_parapsilosis_CDC317_version_s01-m06-r03.gff\",\n",
    "                 \"gff_files/Candida_tropicalis.GCA000006335v3.59.gff3\",\n",
    "                 \"gff_files/Cryptococcus_neoformans.ASM9104v1.59.gff3\",\n",
    "                 \"gff_files/Pichia_kudriavzevii_gca_000764455.ASM76445v1.59.gff3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "23377973-6d75-41ba-877a-bdb55c6f203a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aspergillus fumigatus\n",
      "AFUA_4G06890\n",
      "AFUA_6G05140\n",
      "AFUA_2G00320\n",
      "AFUA_6G12400\n",
      "AFUA_5G05460\n",
      "AFUA_1G05050\n",
      "AFUA_2G15130\n",
      "AFUA_4G12560\n",
      "AFUA_2G03700\n",
      "AFUA_3G05760\n",
      "AFUA_5G06070\n",
      "AFUA_1G10910\n",
      "AFUA_5G10370\n",
      "AFUA_7G03740\n",
      "Candida albicans\n",
      "C5_00660C_A\n",
      "C1_04770C_A\n",
      "C1_02420C_A\n",
      "CR_00850C_A\n",
      "C5_03390C_A\n",
      "C6_00620W_A\n",
      "C3_05920W_A\n",
      "C3_05220W_A\n",
      "C5_01840C_A\n",
      "C1_08460C_A\n",
      "C3_04890W_A\n",
      "C3_07860C_A\n",
      "C6_03170C_A\n",
      "C1_08590C_A\n",
      "C3_02220W_A\n",
      "C1_00800C_A\n",
      "C1_03780C_A\n",
      "C3_06850W_A\n",
      "CR_08780W_A\n",
      "CR_08800W_A\n",
      "C1_00710C_A\n",
      "Candida auris\n",
      "B9J08_001448\n",
      "B9J08_003737\n",
      "B9J08_000964\n",
      "B9J08_001020\n",
      "B9J08_004076\n",
      "B9J08_005397\n",
      "B9J08_000164\n",
      "B9J08_004061\n",
      "B9J08_004819\n",
      "B9J08_000270\n",
      "B9J08_003224\n",
      "B9J08_001232\n",
      "B9J08_001055.75\n",
      "B9J08_001055.25\n",
      "B9J08_003245\n",
      "Nasakeomyces glabrata\n",
      "CAGL0E04334g\n",
      "CAGL0F01793g\n",
      "CAGL0G01034g\n",
      "CAGL0K04037g\n",
      "CAGL0H09064g\n",
      "CAGL0D01562g\n",
      "CAGL0A00451g\n",
      "CAGL0I07733g\n",
      "CAGL0M01760g\n",
      "CAGL0I07755g\n",
      "CAGL0F07865g\n",
      "CAGL0L11506g\n",
      "CAGL0L13392r\n",
      "CAGL0L13376r\n",
      "CAGL0K12650g\n",
      "Candida parapsilosis\n",
      "CPAR2_303740\n",
      "CPAR2_105550\n",
      "CPAR2_106400\n",
      "CPAR2_804030\n",
      "CPAR2_502030\n",
      "CPAR2_602820\n",
      "CPAR2_405290\n",
      "CPAR2_807270\n",
      "CPAR2_303510\n",
      "CPAR2_207280\n",
      "CPAR2_110330\n",
      "CPAR2_808620\n",
      "CPAR2_701565\n",
      "CPAR2_701550\n",
      "CPAR2_105880\n",
      "Candida tropicalis\n",
      "CTRG_05283\n",
      "CTRG_04480\n",
      "CTRG_04661\n",
      "CTRG_02689\n",
      "CTRG_02606\n",
      "CTRG_02269\n",
      "CTRG_05307\n",
      "CTRG_03159\n",
      "CTRG_04347\n",
      "CTRG_05664\n",
      "CTRG_02412\n",
      "Cryptococcus neoformans\n",
      "CNA00300\n",
      "CNN02320\n",
      "CNE02100\n",
      "CNA05950\n",
      "CNF04830\n",
      "CNB04020\n",
      "CNC01920\n",
      "CNA07090\n",
      "Pichia kudriavzevii\n",
      "JL09_g2508\n",
      "JL09_g3074\n",
      "JL09_g1956\n",
      "JL09_g2004\n",
      "JL09_g66\n",
      "JL09_g570\n"
     ]
    }
   ],
   "source": [
    "synth = \"\"\n",
    "spacer = \"NNNNNNNNN\"\n",
    "\n",
    "for i, file in enumerate(species_genome_file_list):\n",
    "    species = species_list[i]\n",
    "    print(species)\n",
    "    with open(file) as genome:\n",
    "        genome_content = genome.read()\n",
    "    \n",
    "    # Removes first line puts a space\n",
    "    dna = re.sub(\">.*\\\\n\", \" \", genome_content)\n",
    "    \n",
    "    gff_file = gff_files_list[i]\n",
    "    with open(gff_file) as gff:\n",
    "        gff_content = gff.read()\n",
    "    \n",
    "    gene_list = df['Gene ID'][df['Species'] == species].values[0]\n",
    "    \n",
    "    for g in gene_list:\n",
    "        print(g)\n",
    "        \n",
    "        # find gene coordinates\n",
    "        y = re.findall(f\"gene\\\\t.*?{g};\", gff_content)\n",
    "        #print(f\"Found: {y}\")\n",
    "        coordinates = re.findall(\"[0-9]+\", y[0])\n",
    "        #print(coordinates)\n",
    "        begin = int(coordinates[0])\n",
    "        end = int(coordinates[1])\n",
    "        \n",
    "        gene_coor = dna[begin:end]\n",
    "        gene = re.sub(\"\\\\n\", \"\", gene_coor)\n",
    "        synth += gene + spacer + \"W\"\n",
    "    \n",
    "    synth += spacer + spacer + \"S\"        \n",
    "\n",
    "my_synth_genome = open(\"my_synthetic_genome.txt\", \"w\")\n",
    "my_synth_genome.write(synth)\n",
    "my_synth_genome.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31af142a-2c3b-4588-a6a6-80dac509282b",
   "metadata": {},
   "source": [
    "### It works but some problems: \n",
    "\n",
    "Candida albicans  and tropicalis both have 1 unrecognisable genes. Respectively C1_00710C_A and CTRG_02689. *Solved !* Problem was erronous \" \" after Gene ID in xlsx file.\n",
    "\n",
    "Candida tropicalis' gff file is broken. *Solved !* I changed the gff and fasta files. Took the top level fasta and the \"normal\" gff3\n",
    "\n",
    "And are they finding the right sequences.... I'm not sure. Large NNN sequence in Cryptococcus ?? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03cbf4bc-24fd-4ace-83c3-cf1f15e1fd2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
